File: mcp-servers/adeotek-mcp-stdio-sql/src/security/queryValidator.ts
/**
 * Query validation and security layer
 * Implements multiple layers of protection to ensure read-only access
 */

import { QueryValidationError } from '../utils/errors.js';
import { ValidationResult } from '../types/index.js';
import logger from '../utils/logger.js';

/**
 * List of dangerous SQL keywords that modify data or schema
 */
const BLOCKED_KEYWORDS = [
  // Data modification
  'INSERT',
  'UPDATE',
  'DELETE',
  'TRUNCATE',
  'MERGE',
  'UPSERT',
  'REPLACE',
  'COPY',
  // Schema modification
  'CREATE',
  'ALTER',
  'DROP',
  'RENAME',
  'COMMENT',
  // Permissions
  'GRANT',
  'REVOKE',
  // Transaction control
  'BEGIN',
  'COMMIT',
  'ROLLBACK',
  'SAVEPOINT',
  'START TRANSACTION',
  // Locking
  'LOCK',
  'UNLOCK',
  // Maintenance
  'VACUUM',
  'ANALYZE',
  'REINDEX',
  'CLUSTER',
  'CHECKPOINT',
  // Configuration
  'SET',
  'RESET',
  // Messaging
  'LISTEN',
  'NOTIFY',
  'UNLISTEN',
  // Procedural
  'DO',
  'CALL',
  'EXECUTE',
  'EXEC',
  'DECLARE',
  // Data definition
  'INDEX',
  'SEQUENCE',
  'VIEW',
  'TRIGGER',
  'FUNCTION',
  'PROCEDURE',
  'SCHEMA',
  'DATABASE',
  'TABLE',
  'CONSTRAINT',
];

/**
 * Dangerous SQL functions that could be used for unauthorized access
 */
const BLOCKED_FUNCTIONS = [
  'pg_read_file',
  'pg_read_binary_file',
  'pg_execute',
  'pg_terminate_backend',
  'pg_cancel_backend',
  'pg_sleep',
  'pg_reload_conf',
  'pg_rotate_logfile',
  'xp_cmdshell', // SQL Server
  'sp_executesql', // SQL Server
  'OPENROWSET', // SQL Server
  'OPENDATASOURCE', // SQL Server
];

/**
 * Pattern-based validation rules
 */
const DANGEROUS_PATTERNS = [
  /;\s*(INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|TRUNCATE|GRANT|REVOKE)/i,
  /--.*?(INSERT|UPDATE|DELETE|DROP|CREATE|ALTER)/i,
  /\/\*.*?(INSERT|UPDATE|DELETE|DROP|CREATE|ALTER).*?\*\//i,
  /INTO\s+OUTFILE/i,
  /LOAD_FILE/i,
  /\$\$/i, // PostgreSQL procedural code blocks
];

/**
 * Validate a SQL query for read-only compliance
 */
export function validateQuery(query: string, maxQueryLength = 50000): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // Basic validation
  if (!query || query.trim().length === 0) {
    errors.push('Query cannot be empty');
    return { isValid: false, errors, warnings };
  }

  if (query.length > maxQueryLength) {
    errors.push(`Query exceeds maximum length of ${maxQueryLength} characters`);
    return { isValid: false, errors, warnings };
  }

  // Normalize query for analysis (preserve original for execution)
  const normalizedQuery = query.trim().toUpperCase();

  // Check if query starts with SELECT, WITH, or EXPLAIN
  const allowedStarters = ['SELECT', 'WITH', 'EXPLAIN', 'SHOW', 'DESCRIBE', 'DESC'];
  const startsWithAllowed = allowedStarters.some((starter) =>
    normalizedQuery.startsWith(starter)
  );

  if (!startsWithAllowed) {
    errors.push(
      `Query must start with one of: ${allowedStarters.join(', ')}. Got: ${normalizedQuery.split(' ')[0] || 'empty'}`
    );
  }

  // Check for blocked keywords
  for (const keyword of BLOCKED_KEYWORDS) {
    const regex = new RegExp(`\\b${keyword}\\b`, 'i');
    if (regex.test(query)) {
      errors.push(`Blocked keyword detected: ${keyword}`);
    }
  }

  // Check for blocked functions
  for (const func of BLOCKED_FUNCTIONS) {
    const regex = new RegExp(`\\b${func}\\s*\\(`, 'i');
    if (regex.test(query)) {
      errors.push(`Blocked function detected: ${func}`);
    }
  }

  // Check for dangerous patterns
  for (const pattern of DANGEROUS_PATTERNS) {
    if (pattern.test(query)) {
      errors.push(`Dangerous SQL pattern detected: ${pattern.toString()}`);
    }
  }

  // Check for multiple statements (semicolons)
  const statementCount = query.split(';').filter((s) => s.trim().length > 0).length;
  if (statementCount > 1) {
    errors.push('Multiple SQL statements are not allowed');
  }

  // Warnings for potentially problematic queries
  if (!/LIMIT\s+\d+/i.test(query) && normalizedQuery.startsWith('SELECT')) {
    warnings.push('Query does not include a LIMIT clause - results may be truncated');
  }

  if (/SELECT\s+\*/i.test(query)) {
    warnings.push('Using SELECT * may retrieve more data than necessary');
  }

  // Log validation result
  if (errors.length > 0) {
    logger.warn('Query validation failed', {
      errors,
      queryPreview: query.substring(0, 100),
    });
  }

  return {
    isValid: errors.length === 0,
    errors,
    warnings,
  };
}

/**
 * Enforce query execution limits
 */
export function enforceQueryLimits(
  query: string,
  maxRows: number = 10000
): { query: string; limitApplied: boolean } {
  // Check if query already has a LIMIT clause
  const limitMatch = query.match(/LIMIT\s+(\d+)/i);

  if (limitMatch) {
    const existingLimit = parseInt(limitMatch[1], 10);
    if (existingLimit <= maxRows) {
      return { query, limitApplied: false };
    }
    // Replace with max allowed limit
    return {
      query: query.replace(/LIMIT\s+\d+/i, `LIMIT ${maxRows}`),
      limitApplied: true,
    };
  }

  // Add LIMIT clause if not present and query is a SELECT
  if (/^\s*(SELECT|WITH)/i.test(query)) {
    return {
      query: `${query.trim()} LIMIT ${maxRows}`,
      limitApplied: true,
    };
  }

  return { query, limitApplied: false };
}

/**
 * Sanitize database/schema/table names to prevent SQL injection
 */
export function sanitizeIdentifier(identifier: string): string {
  // Remove any non-alphanumeric characters except underscore and dot
  const sanitized = identifier.replace(/[^a-zA-Z0-9_.]/g, '');

  if (sanitized !== identifier) {
    logger.warn('Identifier was sanitized', { original: identifier, sanitized });
  }

  return sanitized;
}

/**
 * Validate and throw if query is not safe
 */
export function validateQueryOrThrow(query: string, maxQueryLength?: number): void {
  const result = validateQuery(query, maxQueryLength);

  if (!result.isValid) {
    throw new QueryValidationError(
      `Query validation failed: ${result.errors.join(', ')}`,
      result.errors
    );
  }

  // Log warnings
  if (result.warnings.length > 0) {
    logger.warn('Query validation warnings', { warnings: result.warnings });
  }
}

File: mcp-servers/adeotek-sql-net-mcp/src/AdeotekSqlMcp/Security/QueryValidator.cs
using System.Text.RegularExpressions;
using AdeotekSqlMcp.Utilities;

namespace AdeotekSqlMcp.Security;

/// <summary>
/// Query validation result
/// </summary>
public sealed record ValidationResult
{
    public required bool IsValid { get; init; }
    public IReadOnlyList<string> Errors { get; init; } = Array.Empty<string>();
    public IReadOnlyList<string> Warnings { get; init; } = Array.Empty<string>();
}

/// <summary>
/// Multi-layer SQL query validation for read-only operations
/// </summary>
public sealed class QueryValidator
{
    private static readonly HashSet<string> BlockedKeywords = new(StringComparer.OrdinalIgnoreCase)
    {
        // Data modification
        "INSERT", "UPDATE", "DELETE", "TRUNCATE", "MERGE", "UPSERT", "REPLACE", "COPY",

        // Schema modification
        "CREATE", "ALTER", "DROP", "RENAME", "COMMENT",

        // Permissions
        "GRANT", "REVOKE",

        // Transaction control
        "BEGIN", "COMMIT", "ROLLBACK", "SAVEPOINT", "START TRANSACTION",

        // Locking
        "LOCK", "UNLOCK",

        // Maintenance
        "VACUUM", "ANALYZE", "REINDEX", "CLUSTER", "CHECKPOINT",

        // Configuration
        "SET", "RESET",

        // Messaging
        "LISTEN", "NOTIFY", "UNLISTEN",

        // Procedural
        "DO", "CALL", "EXECUTE", "EXEC", "DECLARE"
    };

    private static readonly HashSet<string> DangerousFunctions = new(StringComparer.OrdinalIgnoreCase)
    {
        // PostgreSQL dangerous functions
        "pg_read_file", "pg_read_binary_file", "pg_execute", "pg_terminate_backend",
        "pg_sleep", "pg_ls_dir", "pg_stat_file",

        // SQL Server dangerous functions
        "xp_cmdshell", "sp_executesql", "OPENROWSET", "OPENDATASOURCE", "OPENQUERY",
        "xp_regread", "xp_regwrite", "xp_fileexist"
    };

    private const int MaxQueryLength = 50000;
    private const int MaxRowLimit = 10000;

    /// <summary>
    /// Validates a SQL query for read-only operations
    /// </summary>
    public ValidationResult Validate(string query)
    {
        var errors = new List<string>();
        var warnings = new List<string>();

        // Check for empty query
        if (string.IsNullOrWhiteSpace(query))
        {
            errors.Add("Query cannot be empty");
            return new ValidationResult { IsValid = false, Errors = errors, Warnings = warnings };
        }

        // Check query length
        if (query.Length > MaxQueryLength)
        {
            errors.Add($"Query exceeds maximum length of {MaxQueryLength} characters");
        }

        // Normalize query
        var normalizedQuery = query.Trim();
        var upperQuery = normalizedQuery.ToUpperInvariant();

        // Check starting keyword (must be SELECT, WITH, or EXPLAIN)
        if (!upperQuery.StartsWith("SELECT") &&
            !upperQuery.StartsWith("WITH") &&
            !upperQuery.StartsWith("EXPLAIN"))
        {
            errors.Add("Query must start with SELECT, WITH, or EXPLAIN");
        }

        // Check for blocked keywords
        foreach (var keyword in BlockedKeywords)
        {
            if (ContainsKeyword(upperQuery, keyword))
            {
                errors.Add($"Blocked keyword detected: {keyword}");
            }
        }

        // Check for dangerous functions
        foreach (var function in DangerousFunctions)
        {
            if (ContainsFunction(upperQuery, function))
            {
                errors.Add($"Dangerous function detected: {function}");
            }
        }

        // Check for multiple statements (semicolon)
        if (ContainsMultipleStatements(normalizedQuery))
        {
            errors.Add("Multiple statements are not allowed");
        }

        // Check for procedural code blocks
        if (upperQuery.Contains("$$") || upperQuery.Contains("$BODY$"))
        {
            errors.Add("Procedural code blocks are not allowed");
        }

        // Check for INTO OUTFILE
        if (Regex.IsMatch(upperQuery, @"\bINTO\s+OUTFILE\b"))
        {
            errors.Add("INTO OUTFILE is not allowed");
        }

        // Check for LOAD_FILE
        if (Regex.IsMatch(upperQuery, @"\bLOAD_FILE\b"))
        {
            errors.Add("LOAD_FILE is not allowed");
        }

        // Check for SQL injection patterns
        if (ContainsSqlInjectionPatterns(normalizedQuery))
        {
            errors.Add("Potential SQL injection pattern detected");
        }

        // Warnings
        if (!Regex.IsMatch(upperQuery, @"\bLIMIT\b") && !Regex.IsMatch(upperQuery, @"\bTOP\b"))
        {
            warnings.Add($"Query does not have a LIMIT clause. Results will be limited to {MaxRowLimit} rows");
        }

        if (upperQuery.Contains("SELECT *"))
        {
            warnings.Add("Using SELECT * may return more data than needed. Consider specifying columns explicitly");
        }

        return new ValidationResult
        {
            IsValid = errors.Count == 0,
            Errors = errors,
            Warnings = warnings
        };
    }

    /// <summary>
    /// Validates query and throws exception if invalid
    /// </summary>
    public void ValidateOrThrow(string query)
    {
        var result = Validate(query);
        if (!result.IsValid)
        {
            var errorMessage = string.Join("; ", result.Errors);
            throw new QueryValidationException($"Query validation failed: {errorMessage}");
        }
    }

    /// <summary>
    /// Enforces row limit on query
    /// </summary>
    public string EnforceLimit(string query, int maxRows = MaxRowLimit)
    {
        maxRows = Math.Min(maxRows, MaxRowLimit);

        var upperQuery = query.ToUpperInvariant();

        // If already has LIMIT, ensure it's not too high
        if (Regex.IsMatch(upperQuery, @"\bLIMIT\s+(\d+)"))
        {
            return Regex.Replace(
                query,
                @"LIMIT\s+(\d+)",
                match =>
                {
                    var limit = int.Parse(match.Groups[1].Value);
                    return limit > maxRows ? $"LIMIT {maxRows}" : match.Value;
                },
                RegexOptions.IgnoreCase
            );
        }

        // If has TOP, ensure it's not too high (SQL Server)
        if (Regex.IsMatch(upperQuery, @"\bTOP\s+(\d+)"))
        {
            return Regex.Replace(
                query,
                @"TOP\s+(\d+)",
                match =>
                {
                    var limit = int.Parse(match.Groups[1].Value);
                    return limit > maxRows ? $"TOP {maxRows}" : match.Value;
                },
                RegexOptions.IgnoreCase
            );
        }

        // Add LIMIT for PostgreSQL-style queries
        if (!upperQuery.Contains("TOP"))
        {
            return $"{query.TrimEnd(';')} LIMIT {maxRows}";
        }

        return query;
    }

    /// <summary>
    /// Sanitizes identifier (table, column, schema names)
    /// </summary>
    public string SanitizeIdentifier(string identifier)
    {
        if (string.IsNullOrWhiteSpace(identifier))
        {
            throw new ArgumentException("Identifier cannot be empty", nameof(identifier));
        }

        // Allow only alphanumeric, underscore, and dot
        var sanitized = Regex.Replace(identifier, @"[^a-zA-Z0-9_.]", "");

        if (sanitized != identifier)
        {
            throw new QueryValidationException($"Invalid identifier: {identifier}");
        }

        return sanitized;
    }

    private static bool ContainsKeyword(string query, string keyword)
    {
        // Use word boundary to avoid false positives (e.g., "inserted" shouldn't match "INSERT")
        var pattern = $@"\b{Regex.Escape(keyword)}\b";
        return Regex.IsMatch(query, pattern, RegexOptions.IgnoreCase);
    }

    private static bool ContainsFunction(string query, string function)
    {
        var pattern = $@"\b{Regex.Escape(function)}\s*\(";
        return Regex.IsMatch(query, pattern, RegexOptions.IgnoreCase);
    }

    private static bool ContainsMultipleStatements(string query)
    {
        // Remove string literals to avoid false positives
        var withoutStrings = Regex.Replace(query, @"'[^']*'", "");

        // Check for semicolons (multiple statements)
        var semicolonCount = withoutStrings.Count(c => c == ';');

        // Allow one trailing semicolon
        if (semicolonCount > 1 || (semicolonCount == 1 && !withoutStrings.TrimEnd().EndsWith(";")))
        {
            return true;
        }

        return false;
    }

    private static bool ContainsSqlInjectionPatterns(string query)
    {
        // Common SQL injection patterns
        var patterns = new[]
        {
            @"'\s*OR\s+'1'\s*=\s*'1",
            @"'\s*OR\s+1\s*=\s*1",
            @"--\s*$",
            @"/\*.*\*/",
            @"\bUNION\s+ALL\s+SELECT\b",
            @"'\s*;\s*DROP\b"
        };

        return patterns.Any(pattern => Regex.IsMatch(query, pattern, RegexOptions.IgnoreCase));
    }
}
